src/server/_security.py

```diff
@@ -0,0 +1,314 @@
+    "arch",
+]
+
+
+class User:
+    api_key: str
+    email: str
+    authenticated: bool
+    roles: Set[UserRole]
+    tracking: bool = True
+    registered: bool = True
+
+    def __init__(self, api_key: str, email: str, authenticated: bool, roles: Set[UserRole], tracking: bool = True, registered: bool = False) -> None:
+        self.api_key = api_key
+        self.email = email
+        self.authenticated = authenticated
+        self.roles = roles
+        self.tracking = tracking
+        self.registered = registered
+
+    def get_apikey(self) -> str:
+        return self.api_key
+
+    def is_authenticated(self) -> bool:
+        return self.authenticated
+
+    def has_role(self, role: UserRole) -> bool:
+        return role in self.roles
+
+    def is_rate_limited(self) -> bool:
+        return not self.registered
+
+    def is_tracking(self) -> bool:
+        return self.tracking
+
+    def log_info(self, msg: str, *args, **kwargs) -> None:
+        # Use structured logger instead of base logger
+        logger = get_structured_logger("api_key_logs", filename="api_key_logs.log")
+        if self.is_authenticated():
+            if self.is_tracking():
+                logger.info(msg, *args, **dict(kwargs, apikey=self.get_apikey()))
+            else:
+                logger.info(msg, *args, **dict(kwargs, apikey="*****"))
+        else:
+            # app.logger.info(msg, *args, **kwargs)
+            logger.info(msg, *args, **kwargs)
+
+
+ANONYMOUS_USER = User("anonymous", "", False, set())
+
+
+def _find_user(api_key: Optional[str]) -> User:
+    if not api_key:
+        return ANONYMOUS_USER
+    stmt = user_table.select().where(user_table.c.api_key == api_key)
+    user = db.execution_options(stream_results=False).execute(stmt).first()
+    if user is None:
+        return ANONYMOUS_USER
+    else:
+        return User(user.api_key, user.email, True, set(user.roles.split(",")), user.tracking, user.registered)
+
+
+def resolve_auth_token() -> Optional[str]:
+    for name in ('auth', 'api_key', 'token'):
+        if name in request.values:
+            return request.values[name]
+    # user name password
+    if request.authorization and request.authorization.username == "epidata":
+        return request.authorization.password
+    # bearer token authentication
+    auth_header = request.headers.get("Authorization")
+    if auth_header and auth_header.startswith("Bearer "):
+        return auth_header[len("Bearer "):]
+    return None
+
+
+def _get_current_user() -> User:
+    if "user" not in g:
+        api_key = resolve_auth_token()
+        user = _find_user(api_key)
+        request_path = request.full_path
+        if not user.is_authenticated() and require_api_key():
+            raise MissingAPIKeyException()
+        # If the user configured no-track option, mask the API key
+        if not user.is_tracking():
+            request_path = mask_apikey(request_path)
+        user.log_info("Get path", path=request_path)
```

Comment:
this method is called all over the place, using the LocalProxy'd `current_user` object declared below.  the logging and exception raising should happen in the appropriate places where this method is called, and not be done in this method itself.